### ISSUE TITLE
Add a warning alert when migrating to 1.25 if they have PSPs.

### ISSUE THREAD

**Issue by gaktive** [rancher/dashboard#7553]
Internal reference: SURE-5392
Related backend work: https://github.com/rancher/rancher/issues/38875

From @snasovich:

When v1.25 is introduced in Rancher, the k3s and rke2 upgrades will break. I'm not 100% sure what will happen when upgrading provisioned clusters, but imported clusters will fail due to them using system-upgrade-controller under the hood. This is because hardened k3s/rke2 clusters use PSPs, and PSPs are removed in v1.25. Currently, system-upgrade-controller does not do any mitigation for these.

For k3s, see k3s-io/k3s#6121
For rke2, the steps to do a SUC upgrade to v1.25 are:
 - Change config.yaml on each node to use profile: cis-1.23 instead of profile: cis-1.6
- Run standard system-upgrade-controller steps. This step, from the Rancher side, just means selecting the v1.25 version desired from the dropdown when editing the cluster through the UI.

Describe alternatives you've considered

Lots of warnings everywhere
Update system-upgrade-controller itself or the Plans or the upgrade images to do any workarounds automagically.

Other notes:

For upgrading provisioned rke2, lock the only profile version option to cis-1.23. If there was previously a selection of cis-1.6 or cis-1.5, give the user a warning. For upgrading provisioned k3s, give the user a warning with instructions on which values to select. For upgrading imported rke2, have the user perform the steps mentioned above first on their nodes, then use the Rancher UI to upgrade. For upgrading imported k3s, provide a warning that the upgrade should be performed manually, and do not give any option to upgrade from a <v1.25 cluster to a >=v1.25.

**Comment by jiaqiluo** [comment]
please see my this comment for the backend's behavior: https://github.com/rancher/rancher/issues/38875#issuecomment-1341331587
TLDR: there is no logic in the backend of Rancher to handle the hardening of an RKE2/K3S cluster, including the following three types: node-driver, custom, and imported. 
 
The only place where UI shows the `CIS Profile` option is when creating/updating a custom RKE2 cluster; however it will not work because 1) the backend does not handle it and 2) UI sets `protect-kernel-defaults: false` at the same time but "true" is required by RKE2. 

<img width="1390" alt="image" src="https://user-images.githubusercontent.com/6218999/206263652-3c89d244-b6ab-448a-871e-bb5ce768cd7b.png">

BTW, users can always set the `CIS Profile` option via `Edit as YAML`, but again it does not work as explained above.

**Comment by catherineluse** [comment]
@jiaqiluo I don't think this affects the warning message PR, but just FYI, there has been a UI fix for the issue where the UI was not setting the `protect-kernel-defaults` to `true` when a CIS profile is selected for custom RKE2 clusters. https://github.com/rancher/dashboard/pull/7662

**Comment by markusewalker** [comment]
As this is part of the broader K8s 1.25 internal test plan, this will be tested alongside other test cases. Once tested, I will close this issue.

**Comment by markusewalker** [comment]
As an update, I am currently waiting for https://github.com/rancher/rancher/issues/40030 to be ready to test. There are SUC test cases in the internal test plan that I would like to validate before marking this issue as completed.

**Comment by markusewalker** [comment]
Validated the test cases associated with this issue in the internal test plan, so I will close this issue.

**Comment by gaktive** [comment]
I'm assuming this was meant to be Reopened. @markusewalker correct me if I'm wrong though if you're waiting on backend and the UI portion is done, then I don't know if there's an apt QA column to have this wait in.

**Comment by gaktive** [comment]
Maybe `QA Blocked` if no UI Dev is required at this time.

**Comment by markusewalker** [comment]
@gaktive correct, this is reopened after speaking to @jiaqiluo offline. We noted that with RKE2 custom cluster being version 1.25, the only available profile version should be `cis-1.23`. He makes mention of it in this [comment](https://github.com/rancher/dashboard/issues/7553#issuecomment-1341387929).

So this will be a use case QA will need to test once the backend work is complete.

**Comment by jiaqiluo** [comment]
hi @gaktive, just want to clarify that we still need UI to fix the incorrect options showing when the RKE2 cluster version is 1.25. ( AFAIK, UI populates the option in the "CIS Worker Profile" list, right? ) 
cc @markusewalker

**Comment by gaktive** [comment]
@jiaqiluo the renaming of CIS Profile is now merged and ready to test in https://github.com/rancher/dashboard/issues/7950 -- PR https://github.com/rancher/dashboard/pull/8035 shows the new wording. Does that cover what's asked for here?

**Comment by cwayne18** [comment]
@jiaqiluo @gaktive is that something we could be pulling from KDM? Maybe we can add a new field to track compatible cis-profiles for a given k8s minor so that we don't have to hardcode this for just 1.25.  This will of course have to go along with a larger effort to enable cis-profiles as they come out (e.g. cis-1.24 is released recently but not added here)

**Comment by cnotv** [comment]
@jiaqiluo As mentioned somewhere else, these values are provided by the server and are not hardcoded from the UI.

**Comment by snasovich** [comment]
@gaktive @cnotv @nwmac @torchiaf , there will be a sync meeting to cover some of the UI expectations, but I wanted to share some thoughts here.

For this issue, the title and the description seem very much out of sync, so I'm not quite sure what's the expectation on the desired behavior/change.

I'm going by the title "Add a warning alert when migrating to 1.25 if they have PSPs.". In this case, I see a few issues with the [PR](https://github.com/rancher/dashboard/pull/7821) linked to this issue (please correct me if I'm wrong).
1. `allPSPs` we're checking seem to be “PSP Templates that are defined on the local cluster” per [code](https://github.com/rancher/dashboard/blob/0e33fd4c13ca5eddfbfb92d4a7a7f509849b555c/shell/edit/provisioning.cattle.io.cluster/rke2.vue#L139) while what we need here is to check “PSPs defined in the downstream cluster we're editing”. This may be somewhat challenging to determine (or not feasible to check from UI side).
2. We only want the warning to be presented when upgrade is done from k8s version <1.25 to k8s version >=1.25 which doesn't seem to be covered by the linked PR.

cc: @jiaqiluo @cwayne18 @markusewalker

**Comment by cnotv** [comment]
@jiaqiluo Could you please define the specification and the viability to identify the presence of the PSP in a cluster as mentioned in the point 1 of @snasovich

**Comment by jiaqiluo** [comment]
hi @cnotv,

For the selected downstream cluster whose k8s version is 1.24 or less, if the user changes the k8s version to 1.25 and above in the edit page, UI should show a warning message if there is `PodSecurityPolicy` resource detected in the cluster. 

The `PodSecurityPolicy` resource in the target cluster can be fetched at the following endpoint 
```
https://<URL>/k8s/clusters/<Cluster-ID>/v1/policy.podsecuritypolicies
```

Let me know if you have more questions.

**Comment by cnotv** [comment]
For the records, it seems like we have already an implementation of this type [here](https://github.com/rancher/dashboard/blob/master/shell/pages/c/_cluster/explorer/index.vue#L151-L163).

**Comment by cnotv** [comment]
To recreate the same conditions, should I do the same of what suggested in [this comment](https://github.com/rancher/dashboard/issues/8125#issuecomment-1438931325)?

**Comment by cnotv** [comment]
> The `PodSecurityPolicy` resource in the target cluster can be fetched at the following endpoint
> 
> ```
> https://<URL>/k8s/clusters/<Cluster-ID>/v1/policy.podsecuritypolicies
> ```

@jiaqiluo I get a 500 when I try to do that, even checked for a possible typo since in the code [there is already some logic](https://github.com/rancher/dashboard/blob/master/shell/pages/c/_cluster/explorer/index.vue#L151-L160) with the [plural form](https://github.com/rancher/dashboard/blob/master/shell/config/types.js#L56).

![Screenshot 2023-02-22 at 16 42 32](https://user-images.githubusercontent.com/5009481/220677620-bc35f0b4-2c33-43f9-a70e-e367956d8e82.png)
![Screenshot 2023-02-22 at 16 43 18](https://user-images.githubusercontent.com/5009481/220677628-fbdb98ab-8aee-40f1-95b6-3c9330e3d2fc.png)

**Comment by cnotv** [comment]
Just to recap what has been discussed between @kwwii and @nwmac, we need to update the alert for:
- RKE2 if a template is selected
- K3S if it's [applied the arg](https://github.com/rancher/dashboard/issues/8125#issuecomment-1438931325)
- RKE1 with a template, but it should require a separated ticket

When creating an RKE2 cluster and we select a template, it also create a `PodSecurityPolicy` resource, which can be requested with the API mentioned by @jiaqiluo :

`https://<URL>/k8s/clusters/<MGMT-Cluster-name>/v1/policy.podsecuritypolicies` <-- Note that's the Management Cluster Name

If either any of these checks fails, we agreed to simply rephrase the alert in 
`Pod Security Policies are not available in Kubernetes v1.25. If you upgrade and have any Pod Security Policies active, it will be removed`

Note: In order to generate a PSP, the cluster must be successfully be provisioned and not only created.

**Comment by markusewalker** [comment]
**QA TEST PLAN**

|Scenario|Test Case|
|---|---|
|1|As a standard user, provision a downstream RKE2 v1.24 node driver cluster with PSP set to `restricted`, upgrade to v1.25 and validate that there is a warning banner|
|2|As a standard user, provision a downstream RKE2 v1.24 node driver cluster with PSP set to `restricted`, edit cluster to remove PSP, remove existing PSP objects, upgrade to v1.25 and validate that there is no warning banner|
|3|As a standard user, provision a downstream RKE2 v1.24 node driver cluster with PSP set to `Default - RKE2 Embedded`, edit cluster to remove PSP and to add PSA set to `rancher-privileged`, remove existing PSP objects, upgrade to v1.25 and validate that there is no warning banner|
|4|As a standard user, provision a downstream RKE2 v1.24 node driver cluster with PSP set to `Default - RKE2 Embedded`, edit cluster to remove PSP and to add PSA set to `rancher-privileged`, upgrade to v1.25 and validate that there is a warning banner|
|5|As a standard user, provision a downstream RKE2 v1.23 hardened custom cluster with CIS profile set to `cis-profile-1.6` and PSP set to `Default - RKE2 Embedded`, upgrade to v1.25 and validate that there is a warning banner|

**Comment by markusewalker** [comment]
Validated that this is addressed on `v2.7-head `https://github.com/rancher/rancher/commit/cca09e4768af17067ca3a848a6b3e5ed44e99aa1.

**ENVIRONMENT DETAILS**
- Rancher install: Helm
- Rancher version: `v2.7-head` https://github.com/rancher/rancher/commit/cca09e4768af17067ca3a848a6b3e5ed44e99aa1
- Browser: Chrome

**TEST RESULT**
|Scenario|Test Case|Result
|---|---|---|
|1|As a standard user, provision a downstream RKE2 v1.24 node driver cluster with PSP set to `restricted`, upgrade to v1.25 and validate that there is a warning banner|:white_check_mark:|
|2|As a standard user, provision a downstream RKE2 v1.24 node driver cluster with PSP set to `restricted`, edit cluster to remove PSP, remove existing PSP objects, upgrade to v1.25 and validate that there is no warning banner|:white_check_mark:|
|3|As a standard user, provision a downstream RKE2 v1.24 node driver cluster with PSP set to `Default - RKE2 Embedded`, edit cluster to remove PSP and to add PSA set to `rancher-privileged`, remove existing PSP objects, upgrade to v1.25 and validate that there is no warning banner|:white_check_mark:|
|4|As a standard user, provision a downstream RKE2 v1.24 node driver cluster with PSP set to `Default - RKE2 Embedded`, edit cluster to remove PSP and to add PSA set to `rancher-privileged`, upgrade to v1.25 and validate that there is a warning banner|:white_check_mark:|
|5|As a standard user, provision a downstream RKE2 v1.23 hardened custom cluster with CIS profile set to `cis-profile-1.6` and PSP set to `Default - RKE2 Embedded`, upgrade to v1.25 and validate that there is a warning banner|:white_check_mark:|

**VALIDATION STEPS**

**Scenario 1**
1. Provisioned a downstream RKE2 v1.24 node driver cluster and made sure PSP was set to `restricted`.
2. Validated that the cluster and nodes appear `Active` in Cluster Management and Cluster Explorer.
3. Validated that all pods come up as `Active` in the System project page.
4. Validated that there are PSP objects existing in my cluster:
![image](https://user-images.githubusercontent.com/52591819/221964157-e734d3e5-5076-4803-acf7-39e625aa6032.png) 
5. Edited the cluster's config and updated K8s version to v1.25 and validated that I see a warning banner:
![image](https://user-images.githubusercontent.com/52591819/221964182-e16ddf50-a5ee-49cf-854a-2fc4fdd5fca3.png) 
6. Upgraded the cluster to v1.25; validated that the cluster and nodes successfully upgraded to v1.25 in Cluster Management and Cluster Explorer.
7. Validated that the pods in the System project page are all still `Active`.
8. Validated that there are no PSP objects lingering in the upgraded cluster.
_______
**Scenario 2**
1. Repeated scenario 1 steps 1-4.
2. Removed the existing PSP objects in the cluster.
3. Edited the cluster's config and updated K8s version to v1.25; validated that I do not see a warning banner.
4. Repeated scenario 1 steps 6-8.
_______
**Scenario 3**
1. Repeated scenario 1 steps 1-4, but used `Default -RKE2 Embedded` as the PSP instead of `restricted.
2. Removed the existing PSP objects in the cluster.
3. Edited the cluster's config and updated K8s version to v1.25 and PSA to `rancher-privileged`; validated that I do not see a warning banner.
4. Repeated scenario 1 steps 6-8.
5. Created a test deployment with image `nginx:latest` in the `default` namespace; validated that it created successfully due to `privileged` being the inherited PSS.
_______
**Scenario 4**
1. Repeated scenario 1 steps 1-4, but used `Default - RKE2 Embedded` as the PSP instead of `restricted`.
2. Edited the cluster's config and updated K8s version to v1.25 and PSA to `rancher-privileged`; validated that I see a warning banner:
![image](https://user-images.githubusercontent.com/52591819/221964360-3df2077c-3307-43a5-a1b2-c5d286e42b9c.png) 
3. Repeated scenario 1 steps 6-8.
4. Created a test deployment with image `nginx:latest` in the `default` namespace; validated that it created successfully due to `privileged` being the inherited PSS.
_______
**Scenario 5**
1. Repeated scenario 1 steps 1-4, but used a hardened RKE2 v1.24 downstream custom cluster with `cis-profile-1.6` and PSP set to `Default - RKE2 Embedded`.
2. Installed the CIS benchmark on this chart successfully.
3. Ran a CIS scan and validated that it passed, proving this is a hardened cluster.
4. Edited the cluster's config and updated K8s version to v1.25; validated that I see a warning banner:
![image](https://user-images.githubusercontent.com/52591819/221964416-7da931e3-c7ee-4353-92c8-861471832797.png) 
5. Repeated scenario 1 steps 6-8.
